File: modules/form/src/core/expression.ts
import { getAtPath } from './pointer';
import { isJsonValue, jsonValuesEqual } from './schema-profile';
import type { FormExpression, JsonObject, JsonPrimitive, JsonValue } from './types';

export interface ExpressionOptions {
  maxOperations?: number;
  /** Resolves a declared field-path template for the current evaluation scope. */
  resolveFieldPath?: (path: string) => string | undefined;
}

export interface ExpressionAnalysis {
  size: number;
  fieldPaths: string[];
}

const unaryOperators = new Set(['not', 'exists']);
const collectionOperators = new Set(['all', 'any', 'coalesce', 'concat']);
const binaryOperators = new Set([
  'eq',
  'ne',
  'gt',
  'gte',
  'lt',
  'lte',
  'contains',
  'in',
  'add',
  'subtract',
  'multiply',
  'divide',
]);

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === 'object' && !Array.isArray(value);
}

function assertOnlyKeys(node: Record<string, unknown>, allowed: readonly string[]): void {
  const allowedKeys = new Set(allowed);
  const unexpected = Object.keys(node).find(
    (key) => node[key] !== undefined && !allowedKeys.has(key),
  );
  if (unexpected) throw new TypeError(`Expression contains unexpected property ${unexpected}.`);
}

export function analyzeExpression(input: unknown): ExpressionAnalysis {
  const fieldPaths = new Set<string>();
  let size = 0;
  const visit = (value: unknown): void => {
    if (!isRecord(value) || typeof value.op !== 'string') {
      throw new TypeError('Expression must be an object with a supported operator.');
    }
    size += 1;
    if (value.op === 'literal') {
      assertOnlyKeys(value, ['op', 'value']);
      if (!isJsonValue(value.value))
        throw new TypeError('Expression literal must be a JSON value.');
      return;
    }
    if (value.op === 'field') {
      assertOnlyKeys(value, ['op', 'path']);
      if (
        typeof value.path !== 'string' ||
        value.path.length === 0 ||
        value.path.split('.').some((segment) => segment.length === 0)
      ) {
        throw new TypeError('Expression field path must contain non-empty dot-separated segments.');
      }
      fieldPaths.add(value.path);
      return;
    }
    if (unaryOperators.has(value.op)) {
      assertOnlyKeys(value, ['op', 'value']);
      visit(value.value);
      return;
    }
    if (collectionOperators.has(value.op)) {
      assertOnlyKeys(value, ['op', 'values']);
      if (!Array.isArray(value.values)) {
        throw new TypeError(`Expression ${value.op} values must be an array.`);
      }
      for (const child of value.values) visit(child);
      return;
    }
    if (value.op === 'if') {
      assertOnlyKeys(value, ['op', 'condition', 'whenTrue', 'whenFalse']);
      visit(value.condition);
      visit(value.whenTrue);
      visit(value.whenFalse);
      return;
    }
    if (binaryOperators.has(value.op)) {
      assertOnlyKeys(value, ['op', 'left', 'right']);
      visit(value.left);
      visit(value.right);
      return;
    }
    throw new TypeError(`Unsupported expression operator ${value.op}.`);
  };
  visit(input);
  return { size, fieldPaths: [...fieldPaths] };
}

export function expressionFieldPaths(expression: FormExpression): string[] {
  return analyzeExpression(expression).fieldPaths;
}

function comparable(value: unknown): string | number | boolean | null | undefined {
  return typeof value === 'string' ||
    typeof value === 'number' ||
    typeof value === 'boolean' ||
    value === null
    ? value
    : undefined;
}

function finiteNumber(value: JsonValue | undefined, operator: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value)) {
    throw new TypeError(`Expression ${operator} operands must be finite numbers.`);
  }
  return value;
}

function finiteResult(value: number, operator: string): number {
  if (!Number.isFinite(value)) {
    throw new RangeError(`Expression ${operator} result must be a finite number.`);
  }
  return value;
}

function scalarText(value: JsonValue | undefined): string {
  if (value === undefined || value === null) return '';
  if (typeof value === 'object') {
    throw new TypeError('Expression concat values must be JSON primitives.');
  }
  return String(value as JsonPrimitive);
}

export function evaluateExpression(
  expression: FormExpression,
  value: JsonObject,
  options: ExpressionOptions = {},
): JsonValue | undefined {
  const limit = options.maxOperations ?? 256;
  let operations = 0;
  const evaluate = (node: FormExpression): JsonValue | undefined => {
    operations += 1;
    if (operations > limit) throw new Error(`Expression operation limit exceeded (${limit}).`);
    switch (node.op) {
      case 'literal':
        return node.value;
      case 'field': {
        const path = options.resolveFieldPath ? options.resolveFieldPath(node.path) : node.path;
        if (!path) throw new Error(`Expression field path ${node.path} could not be resolved.`);
        return getAtPath(value, path) as JsonValue | undefined;
      }
      case 'not':
        return !evaluate(node.value);
      case 'exists': {
        const result = evaluate(node.value);
        return result !== undefined && result !== null && result !== '';
      }
      case 'all':
        return node.values.every((item) => Boolean(evaluate(item)));
      case 'any':
        return node.values.some((item) => Boolean(evaluate(item)));
      case 'coalesce': {
        for (const item of node.values) {
          const result = evaluate(item);
          if (result !== undefined && result !== null) return result;
        }
        return undefined;
      }
      case 'concat':
        return node.values.map((item) => scalarText(evaluate(item))).join('');
      case 'if':
        return evaluate(node.condition) ? evaluate(node.whenTrue) : evaluate(node.whenFalse);
      default: {
        const left = evaluate(node.left);
        const right = evaluate(node.right);
        switch (node.op) {
          case 'eq':
            return jsonValuesEqual(left, right);
          case 'ne':
            return !jsonValuesEqual(left, right);
          case 'gt':
            return (comparable(left) as never) > (comparable(right) as never);
          case 'gte':
            return (comparable(left) as never) >= (comparable(right) as never);
          case 'lt':
            return (comparable(left) as never) < (comparable(right) as never);
          case 'lte':
            return (comparable(left) as never) <= (comparable(right) as never);
          case 'contains':
            return typeof left === 'string'
              ? left.includes(String(right ?? ''))
              : Array.isArray(left) && left.some((item) => jsonValuesEqual(item, right));
          case 'in':
            return Array.isArray(right) && right.some((item) => jsonValuesEqual(item, left));
          case 'add':
            return finiteResult(
              finiteNumber(left, node.op) + finiteNumber(right, node.op),
              node.op,
            );
          case 'subtract':
            return finiteResult(
              finiteNumber(left, node.op) - finiteNumber(right, node.op),
              node.op,
            );
          case 'multiply':
            return finiteResult(
              finiteNumber(left, node.op) * finiteNumber(right, node.op),
              node.op,
            );
          case 'divide': {
            const divisor = finiteNumber(right, node.op);
            if (divisor === 0) throw new RangeError('Expression cannot divide by zero.');
            return finiteResult(finiteNumber(left, node.op) / divisor, node.op);
          }
        }
      }
    }
  };
  return evaluate(expression);
}

File: modules/form/crates/a3s-form-core/src/evaluation/expression.rs
use std::cmp::Ordering;

use crate::{canonicalize_value, CanonicalValue};

use super::path::{get_at_path, resolve_value_path_template};

pub(super) fn evaluate_expression(
    expression: &CanonicalValue,
    value: &CanonicalValue,
    max_operations: u64,
    row_indices: Option<&[usize]>,
) -> Result<Option<CanonicalValue>, String> {
    Evaluator {
        value,
        max_operations,
        operations: 0,
        row_indices,
    }
    .evaluate(expression)
}

struct Evaluator<'a> {
    value: &'a CanonicalValue,
    max_operations: u64,
    operations: u64,
    row_indices: Option<&'a [usize]>,
}

impl Evaluator<'_> {
    fn evaluate(&mut self, node: &CanonicalValue) -> Result<Option<CanonicalValue>, String> {
        self.operations += 1;
        if self.operations > self.max_operations {
            return Err(format!(
                "Error: Expression operation limit exceeded ({}).",
                self.max_operations
            ));
        }
        let operator = node
            .get("op")
            .and_then(CanonicalValue::as_str)
            .ok_or_else(|| {
                "TypeError: Expression must be an object with a supported operator.".to_owned()
            })?;
        match operator {
            "literal" => node
                .get("value")
                .cloned()
                .map(Some)
                .ok_or_else(|| "TypeError: Expression literal must be a JSON value.".to_owned()),
            "field" => self.evaluate_field(node),
            "not" => Ok(Some(CanonicalValue::Bool(!truthy(
                self.evaluate_child(node, "value")?.as_ref(),
            )))),
            "exists" => {
                let result = self.evaluate_child(node, "value")?;
                Ok(Some(CanonicalValue::Bool(exists(result.as_ref()))))
            }
            "all" => self.evaluate_all(node),
            "any" => self.evaluate_any(node),
            "coalesce" => self.evaluate_coalesce(node),
            "concat" => self.evaluate_concat(node),
            "if" => {
                let condition = self.evaluate_child(node, "condition")?;
                if truthy(condition.as_ref()) {
                    self.evaluate_child(node, "whenTrue")
                } else {
                    self.evaluate_child(node, "whenFalse")
                }
            }
            "eq" | "ne" | "gt" | "gte" | "lt" | "lte" | "contains" | "in" | "add" | "subtract"
            | "multiply" | "divide" => self.evaluate_binary(node, operator),
            _ => Err(format!(
                "TypeError: Unsupported expression operator {operator}."
            )),
        }
    }

    fn evaluate_field(&self, node: &CanonicalValue) -> Result<Option<CanonicalValue>, String> {
        let declared = node
            .get("path")
            .and_then(CanonicalValue::as_str)
            .ok_or_else(|| {
                "TypeError: Expression field path must contain non-empty dot-separated segments."
                    .to_owned()
            })?;
        let resolved = match self.row_indices {
            Some(indices) => resolve_value_path_template(declared, indices),
            None => Some(declared.to_owned()),
        }
        .ok_or_else(|| format!("Error: Expression field path {declared} could not be resolved."))?;
        Ok(get_at_path(self.value, &resolved).cloned())
    }

    fn evaluate_child(
        &mut self,
        node: &CanonicalValue,
        key: &str,
    ) -> Result<Option<CanonicalValue>, String> {
        let child = node.get(key).ok_or_else(|| {
            "TypeError: Expression must be an object with a supported operator.".to_owned()
        })?;
        self.evaluate(child)
    }

    fn collection<'a>(&self, node: &'a CanonicalValue) -> Result<&'a [CanonicalValue], String> {
        node.get("values")
            .and_then(CanonicalValue::as_array)
            .ok_or_else(|| {
                let operator = node
                    .get("op")
                    .and_then(CanonicalValue::as_str)
                    .unwrap_or("unknown");
                format!("TypeError: Expression {operator} values must be an array.")
            })
    }

    fn evaluate_all(&mut self, node: &CanonicalValue) -> Result<Option<CanonicalValue>, String> {
        for child in self.collection(node)? {
            if !truthy(self.evaluate(child)?.as_ref()) {
                return Ok(Some(CanonicalValue::Bool(false)));
            }
        }
        Ok(Some(CanonicalValue::Bool(true)))
    }

    fn evaluate_any(&mut self, node: &CanonicalValue) -> Result<Option<CanonicalValue>, String> {
        for child in self.collection(node)? {
            if truthy(self.evaluate(child)?.as_ref()) {
                return Ok(Some(CanonicalValue::Bool(true)));
            }
        }
        Ok(Some(CanonicalValue::Bool(false)))
    }

    fn evaluate_coalesce(
        &mut self,
        node: &CanonicalValue,
    ) -> Result<Option<CanonicalValue>, String> {
        for child in self.collection(node)? {
            let result = self.evaluate(child)?;
            if !matches!(result, None | Some(CanonicalValue::Null)) {
                return Ok(result);
            }
        }
        Ok(None)
    }

    fn evaluate_concat(&mut self, node: &CanonicalValue) -> Result<Option<CanonicalValue>, String> {
        let mut output = String::new();
        for child in self.collection(node)? {
            output.push_str(&scalar_text(self.evaluate(child)?.as_ref())?);
        }
        Ok(Some(CanonicalValue::String(output)))
    }

    fn evaluate_binary(
        &mut self,
        node: &CanonicalValue,
        operator: &str,
    ) -> Result<Option<CanonicalValue>, String> {
        let left = self.evaluate_child(node, "left")?;
        let right = self.evaluate_child(node, "right")?;
        let result = match operator {
            "eq" => CanonicalValue::Bool(json_values_equal(left.as_ref(), right.as_ref())),
            "ne" => CanonicalValue::Bool(!json_values_equal(left.as_ref(), right.as_ref())),
            "gt" | "gte" | "lt" | "lte" => {
                CanonicalValue::Bool(relational(left.as_ref(), right.as_ref(), operator))
            }
            "contains" => CanonicalValue::Bool(contains(left.as_ref(), right.as_ref())),
            "in" => CanonicalValue::Bool(in_collection(left.as_ref(), right.as_ref())),
            "add" | "subtract" | "multiply" | "divide" => {
                let left_number = finite_number(left.as_ref(), operator)?;
                let right_number = finite_number(right.as_ref(), operator)?;
                if operator == "divide" && right_number == 0.0 {
                    return Err("RangeError: Expression cannot divide by zero.".to_owned());
                }
                let number = match operator {
                    "add" => left_number + right_number,
                    "subtract" => left_number - right_number,
                    "multiply" => left_number * right_number,
                    "divide" => left_number / right_number,
                    _ => {
                        return Err(format!(
                            "TypeError: Unsupported expression operator {operator}."
                        ));
                    }
                };
                if !number.is_finite() {
                    return Err(format!(
                        "RangeError: Expression {operator} result must be a finite number."
                    ));
                }
                CanonicalValue::Number(number)
            }
            _ => {
                return Err(format!(
                    "TypeError: Unsupported expression operator {operator}."
                ));
            }
        };
        Ok(Some(result))
    }
}

fn truthy(value: Option<&CanonicalValue>) -> bool {
    match value {
        None | Some(CanonicalValue::Null) | Some(CanonicalValue::Bool(false)) => false,
        Some(CanonicalValue::Number(value)) => *value != 0.0,
        Some(CanonicalValue::String(value)) => !value.is_empty(),
        Some(CanonicalValue::Bool(true) | CanonicalValue::Array(_) | CanonicalValue::Object(_)) => {
            true
        }
    }
}

fn exists(value: Option<&CanonicalValue>) -> bool {
    match value {
        None | Some(CanonicalValue::Null) => false,
        Some(CanonicalValue::String(value)) => !value.is_empty(),
        Some(_) => true,
    }
}

fn json_values_equal(left: Option<&CanonicalValue>, right: Option<&CanonicalValue>) -> bool {
    let (Some(left), Some(right)) = (left, right) else {
        return false;
    };
    match (canonicalize_value(left), canonicalize_value(right)) {
        (Ok(left), Ok(right)) => left == right,
        _ => false,
    }
}

pub(super) fn optional_values_equal(
    left: Option<&CanonicalValue>,
    right: Option<&CanonicalValue>,
) -> bool {
    match (left, right) {
        (None, None) => true,
        _ => json_values_equal(left, right),
    }
}

fn finite_number(value: Option<&CanonicalValue>, operator: &str) -> Result<f64, String> {
    match value {
        Some(CanonicalValue::Number(value)) if value.is_finite() => Ok(*value),
        _ => Err(format!(
            "TypeError: Expression {operator} operands must be finite numbers."
        )),
    }
}

fn scalar_text(value: Option<&CanonicalValue>) -> Result<String, String> {
    match value {
        None | Some(CanonicalValue::Null) => Ok(String::new()),
        Some(CanonicalValue::String(value)) => Ok(value.clone()),
        Some(CanonicalValue::Bool(value)) => Ok(value.to_string()),
        Some(CanonicalValue::Number(value)) => Ok(number_text(*value)),
        Some(CanonicalValue::Array(_) | CanonicalValue::Object(_)) => {
            Err("TypeError: Expression concat values must be JSON primitives.".to_owned())
        }
    }
}

fn contains(left: Option<&CanonicalValue>, right: Option<&CanonicalValue>) -> bool {
    match left {
        Some(CanonicalValue::String(left)) => left.contains(&javascript_string(right)),
        Some(CanonicalValue::Array(values)) => values
            .iter()
            .any(|item| json_values_equal(Some(item), right)),
        _ => false,
    }
}

fn in_collection(left: Option<&CanonicalValue>, right: Option<&CanonicalValue>) -> bool {
    match right {
        Some(CanonicalValue::Array(values)) => values
            .iter()
            .any(|item| json_values_equal(Some(item), left)),
        _ => false,
    }
}

fn javascript_string(value: Option<&CanonicalValue>) -> String {
    match value {
        None | Some(CanonicalValue::Null) => String::new(),
        Some(CanonicalValue::String(value)) => value.clone(),
        Some(CanonicalValue::Bool(value)) => value.to_string(),
        Some(CanonicalValue::Number(value)) => number_text(*value),
        Some(CanonicalValue::Object(_)) => "[object Object]".to_owned(),
        Some(CanonicalValue::Array(values)) => values
            .iter()
            .map(|value| javascript_string(Some(value)))
            .collect::<Vec<_>>()
            .join(","),
    }
}

fn number_text(value: f64) -> String {
    if value == 0.0 {
        "0".to_owned()
    } else {
        ryu_js::Buffer::new().format_finite(value).to_owned()
    }
}

fn relational(
    left: Option<&CanonicalValue>,
    right: Option<&CanonicalValue>,
    operator: &str,
) -> bool {
    let ordering = match (comparable(left), comparable(right)) {
        (Comparable::String(left), Comparable::String(right)) => {
            Some(left.encode_utf16().cmp(right.encode_utf16()))
        }
        (left, right) => {
            let left = comparable_number(left);
            let right = comparable_number(right);
            if left.is_nan() || right.is_nan() {
                None
            } else {
                left.partial_cmp(&right)
            }
        }
    };
    match operator {
        "gt" => ordering == Some(Ordering::Greater),
        "gte" => matches!(ordering, Some(Ordering::Greater | Ordering::Equal)),
        "lt" => ordering == Some(Ordering::Less),
        "lte" => matches!(ordering, Some(Ordering::Less | Ordering::Equal)),
        _ => false,
    }
}

enum Comparable<'a> {
    Undefined,
    Null,
    Bool(bool),
    Number(f64),
    String(&'a str),
}

fn comparable(value: Option<&CanonicalValue>) -> Comparable<'_> {
    match value {
        None | Some(CanonicalValue::Array(_) | CanonicalValue::Object(_)) => Comparable::Undefined,
        Some(CanonicalValue::Null) => Comparable::Null,
        Some(CanonicalValue::Bool(value)) => Comparable::Bool(*value),
        Some(CanonicalValue::Number(value)) => Comparable::Number(*value),
        Some(CanonicalValue::String(value)) => Comparable::String(value),
    }
}

fn comparable_number(value: Comparable<'_>) -> f64 {
    match value {
        Comparable::Undefined => f64::NAN,
        Comparable::Null => 0.0,
        Comparable::Bool(value) => u8::from(value) as f64,
        Comparable::Number(value) => value,
        Comparable::String(value) => javascript_number(value),
    }
}

fn javascript_number(value: &str) -> f64 {
    let value = value.trim();
    if value.is_empty() {
        return 0.0;
    }
    match value {
        "Infinity" | "+Infinity" => return f64::INFINITY,
        "-Infinity" => return f64::NEG_INFINITY,
        _ => {}
    }
    if let Some(hex) = value
        .strip_prefix("0x")
        .or_else(|| value.strip_prefix("0X"))
    {
        return u64::from_str_radix(hex, 16)
            .map(|value| value as f64)
            .unwrap_or(f64::NAN);
    }
    if let Some(binary) = value
        .strip_prefix("0b")
        .or_else(|| value.strip_prefix("0B"))
    {
        return u64::from_str_radix(binary, 2)
            .map(|value| value as f64)
            .unwrap_or(f64::NAN);
    }
    if let Some(octal) = value
        .strip_prefix("0o")
        .or_else(|| value.strip_prefix("0O"))
    {
        return u64::from_str_radix(octal, 8)
            .map(|value| value as f64)
            .unwrap_or(f64::NAN);
    }
    value.parse::<f64>().unwrap_or(f64::NAN)
}
